src/librbd/journal/Types.cc

```diff
@@ -384,11 +385,15 @@ void EventEntry::decode(bufferlist::iterator& it) {
   }
 
   boost::apply_visitor(DecodeVisitor(struct_v, it), event);
+  if (struct_v >= 4) {
+    ::decode(timestamp, it);
```

Comment:
New fields always need to be added to the end of the structure to preserve compatibility.  You are correct that by placing the encode/decode in this common location, there is the possibility that an updated ```Event``` will now be creating incompatible streams. I think you need to add this to every event structure.